### ISSUE TITLE
Unable to build 32-bit Hercules using Windows SDK 8.1 or 10

### ISSUE THREAD

**Issue by srorso** [hercules-390/hyperion#102]
Building any Hercules using Windows SDK 8.1 or 10 results in the following error messages, followed by a termination of the build.  Windows SDK 8.1 or 10 are the only SDK's supported by Visual Studio 2015. Because there does not seem to be any "bit-ness" in the SDKs, I believe this will be an issue for 64-bit builds as well, but I cannot confirm.  

>C:\Program Files\Windows Kits\10\include\10.0.10586.0\um\winioctl.h(6519): error C2143: syntax error: missing ':' before 'constant'
1>C:\Program Files\Windows Kits\10\include\10.0.10586.0\um\winioctl.h(6520): error C2143: syntax error: missing ':' before 'constant'
1>C:\Program Files\Windows Kits\10\include\10.0.10586.0\um\winioctl.h(6521): error C2143: syntax error: missing ':' before 'constant'
1>C:\Program Files\Windows Kits\10\include\10.0.10586.0\um\winioctl.h(6521): error C2034: '<alignment member>': type of bit field too small for number of bits
1>NMAKE : fatal error U1077: '"C:\Program Files\Microsoft Visual Studio 14.0\VC\BIN\cl.EXE"' : return code '0x2'
1>  Stop.
1>C:\Program Files\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.MakeFile.Targets(37,5): error MSB3073: The command "makefile.bat RETAIL makefile.msvc 32" exited with code 2.

This issue arises because winioctl.h in both SDKs defines the following union:

typedef union {
       USN_RECORD_COMMON_HEADER Header;
       USN_RECORD_V2 V2;
       USN_RECORD_V3 V3;
       USN_RECORD_V4 V4;
} USN_RECORD_UNION, *PUSN_RECORD_UNION;

The union defines the three layouts of the Update Sequence Number records in the NTFS journal; prior to Windows 8.1, there was only one supported layout named USN_RECORD and no union was defined in winioctl.h.  Builds using older SDK's are not affected as a result.  

When HERC_FLAGS.msvc passes preprocessor variables V1, V2, V3, and V4 into the C environment, expansion of those variables creates syntax errors when the union is compiled.  So it's not VS2015 that uses them, it's the Windows 8.1/10 SDK.  

Commits between 971ebe4 and 6989a0d inclusive can be built using the 8.1/10 SDKs because between those commits, V1, V2, V3, and V4 were not passed to the C environment and winioctl.h compiled cleanly. 

V1, V2, V3, and V4 are defined in makefile.bat, set up and assembled into the Version variable in VERSION.msvc, passed to the C preprocessor via -D flags in HERC_FLAGS.msvc, edit-checked inside the C preprocessor environment in version.h, and edit-checked and if necessary defaulted to zero in dyngui.c.  The defaulting in dyngui.c enabled clean builds when the variables were not included in -D flags in the commit range above.  

I will rename the Hercules V1, V2, V3, and V4 variables to VERSION_SUB1 etc., and I'll be happy to implement an alternative solution or variable name.  

Additional information:

NTFS Change Journal Description

[https://msdn.microsoft.com/en-us/library/windows/desktop/aa363798(v=vs.85).aspx](https://msdn.microsoft.com/en-us/library/windows/desktop/aa363798(v=vs.85).aspx)

Description of USN Journal Record Layout V2 and links to V3 and V4

[https://msdn.microsoft.com/en-us/library/windows/desktop/aa365722(v=vs.85).aspx](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365722(v=vs.85).aspx)

**Comment by srorso** [comment]
Oooop, gotta figure out the *nix build changes too...

**Comment by herc4mac** [comment]
as far as I remember unix builds do not care about V1/V2/V3/V4

they use VERSION and that is defined in configure.ac and generated in config.h

I do not understand why the V1/V2/V3/V4 thing is still around 
it is just the contrary of the KISS ( Keep It Simple ) principle 

see my comment in dyngui.c 

cheers
enrico

**Comment by srorso** [comment]
Hi Enrico:

I'm a newbie, so I must assume that the stuff I don't understand is the result of being a newbie.  Happy to make that assumption; without it I won't learn new things, which is why I'm here.  

It won't do me any harm and may do me some good to build Hercules in a VirtualBox running Debian Jessie (8.4) to see how Version/V1/V2/V3/V4 happen in a different environment.  One more introduction to working on an app that runs on lots of platforms.  

Best Regards,
Steve

**Comment by srorso** [comment]
I have done some research.  Reference sources are the commit summary at the end of this post.   

Shorthand alert: in the text below, when you see "Vx Flags," mentally substitute "preprocessor variables V1, V2, V3, and V4."  Lest the detail of correctness obscure the semantics.  

1) The Vx flags exist only in the Windows build environment.  They do not exist in any other build environment (which all appear to be based on configure.ac).
2) The Vx flags (numeric) are required by the .rc/.rc2 files compiled by the Microsoft Resource Compiler.  In Hercules built on/for Windows, compiled resources are used to associate icons with the executables.  (It might be fun to improve the icon associations; right now the Hercules executable gets the H and everything else gets the disk pack--but improvement is not a priority.)
3) In December 2010, the Vx flags were passed to the C compiler environment so that dyngui.c could be assured of a numeric V1 & V2.
4) In March of 2012, VERSION replaced the Vx flags in dyngui.c.  While preprocessor code continues to reference Vx Flags, this is the end of references to them in code processed by the C compiler.  

Since Mar-2012, the presence of the Vx flags in the C compiler environment appear to have been a source of difficulty for builders, some of whom point this out, with no value add in the compiled code.  I'll bet there are others which say, "doesn't work, don't have time/energy to figure it out."  So the Hercules community loses potential developers/enthusiasts. I was in this group last year.  

My conclusions:

1) The Vx flags are required--and therefore add value--to the Windows Resource Compiler.  So no changes to rcflags in HERC_FLAGS.msvc, nor any changes to the preceding files (e.g., makefile.bat) that mess with VxFlags. 
2) Since March of 2012, the Vx flags have not been required in the C Compiler environment, have created problems for Windows and *nix builders from source, and therefore do not add value.  

My path forward (subject to your critique; better ideas welcomed):

1) The Vx flags will not be passed to the C compiler by changing HERC_FLAGS.msvc to exclude -D options for the Vx Flags from cflags.
2) All other Windows build environment components will remain unchanged.
3) .c and .h files referencing the Vx flags will be changed to remove those references (version.h, dyngui.c--sorry Fish and Enrico, the comments should go too). 
4) Initial reference removal in .c and .h files will be by preprocessor directives of the form #if 0/#endif, with ultimate deletion from source in two months time (I welcome suggestions on alternate timeframes). 
5) Initial reference removal in .msvc files will be by commenting with "::" and using the same timeframe for permanent deletion.

Observation/opportunity:

In the Windows enviroment, there's a pretty tight integration between the GitHub repository and the generated VERSION string.  This has been a real help to me as I work with Hercules builds, allowing me to confirm in log files that I'm on the right branch and mod.  It would be real cool to modify configure.ac / autogen.sh so that the VERSION string includes this information.  

Here's a summary of commits relative to the Vx flags:

```
---Date----  Commit   Description
28-Dec-2006, 2634162, HERC_FLAGS.msvc module creation; Vx flags passed to rc but not to c compiler
11-May-2010, 349f9fb, Vx flags used to create VERSION if VERSION not defined in version.h
08-Dec-2010, 9757196, Vx flags passed to c compiler in HERC_FLAGS.msvc
                             Vx flag edits added to version.h, #error if not defined
08-Dec-2010, 1e85707, Vx flag edits in version.h scoped to _MSVC_ only.  
22-Dec-2010, 99ffb70, V1 & V2 introduced to dyngui.c
28-Aug-2011, 85beb7e, V1 & V2 Validation/defaulting code added to dyngui.c,likely to enable *nix builds
10-Mar-2012, a3144fa, V1 & V2 replaced by VERSION in dyngui.c
14-Mar-2016, 40878bc, Vx flags excluded from c compiler in HERC_FLAGS.msvc
20-Apr-2016, 105b9b5, Vx flags passed to c compiler, edits added, in HERC_FLAGS.msvc "Temporary workaround"
                             Vx flag edits scoped to MS VC versions

**Comment by srorso** [comment]
Fix validated by successful builds in Windows VS2015 and in Debian 8.4.

**Comment by Fish-Git** [comment]
> 2) The Vx flags (numeric) are required by the .rc/.rc2 files compiled by the Microsoft Resource Compiler. In Hercules built on/for Windows, compiled resources are used to associate icons with the executables.

Icons are nice, but the _primary_ purpose for resources is to record the product's (or file's) description, version and copyright information (as well as other information too).  That way you don't have to actually execute the program to learn the program's purpose or which version it is. (What command line switches do you use to display that information? What if simply the running the program causes unwanted side effects? Displaying the binary's "Properties" without actually running the program is faster and easier and much safer.)

> 4) In March of 2012, VERSION replaced the Vx flags in dyngui.c. While preprocessor code continues to reference Vx Flags, this is the end of references to them in code processed by the C compiler

Which was a mistake on Enrico's part that I failed to catch. `dyngui.c` _cannot_ simply use `VERSION`. Rather, it _must_ continue to use `V1` and `V2` (but renamed to something else of course).

The GUI expects a floating point number of the form "number-dot-number" and not a string "a.b.c.d-gxxxx-modified". What needs to be done (and I will be committing this change shortly) is to simply _rename_ the `V1` et al. flags to something else and use them instead. (I chose `VERS_MAJ`, `VERS_INT`, VERS_MIN` and `VERS_BLD`, because that's what they really are after all.)

> 2) Since March of 2012, the Vx flags have not been required in the C Compiler environment, have created problems for Windows and *nix builders from source, and therefore do not add value.

As already stated they _do_ add value because they are still very much needed by `dyngui.c` (Windows HercGUI) and need to stay (but renamed of course). Whether other external GUIs also use the Hercules version information that `dyngui` is passing to them, I don't know.

As far as the problem they've caused (as documented by Enrico in his comments in `dyngui.c`, i.e. '08' and '09' being flagged by the C compiler as invalid octal constants), such problems can be easily overcome with minimal effort via a few lines of code in `makefile.bat` (which my upcoming commit does).

How best to deal with the issue on the Nix side I'm not sure, but for now I'm using `#if defined( _MSVC_ ), #else, #endif` around the problematic statement.  I believe this should be good enough for now until an equivalent to `makefile.bat` can be coded for Nix platforms (**_IF_** we actually want to go that route, which is something we should probably discuss elsewhere).

> In the Windows enviroment, there's a pretty tight integration between the GitHub repository and the generated VERSION string. This has been a real help to me as I work with Hercules builds, allowing me to confirm in log files that I'm on the right branch and mod. It would be real cool to modify configure.ac / autogen.sh so that the VERSION string includes this information.

I agree 100% !!!

I started looking into this issue a long time ago but got sidetracked and never got back to it. Somebody should create a new GitHub Issue for this so we can not only track it but also to allow others to offer their own suggestions regarding how best to accomplish it. Some of our users are quite smart after all. You, Steve, are in fact a good example of this. &nbsp; `:)`

In closing I would like to **_THANK YOU_** sincerely, Steve, for looking into and addressing this issue. I am embarrassed to admit I never bothered to examine `winioctl.h` to see what the root cause of the error was. I simply presumed it had something to do with either the Windows 7.1 SDK or `<win32.mak>`, especially given all the other problems I was having with VS2015 at the time.

(Update 1 has _serious_ problems trying to invoke batch files from the IDE, and Update 2, while fixing the Update 1 problems, [introduced new problems of its own](https://connect.microsoft.com/VisualStudio/feedback/details/2614554/build-and-rebuild-solution-fail-to-invoke-defined-makefile-commands))

When I saw you open this issue with your analysis as to the actual cause I was flabbergasted and immediately encouraged to try again getting VS2015 working with Hercules. In the end, thanks to you, I was successful.

We can now successfully build Hercules with VS2015! (although you have to use `makefile.bat` from the command-line; [building directly from the IDE still has problems](https://connect.microsoft.com/VisualStudio/feedback/details/2614554/build-and-rebuild-solution-fail-to-invoke-defined-makefile-commands)).

Great job, Steve, and thanks again!

**Comment by Fish-Git** [comment]
@herc4mac wrote:

> I do not understand why the V1/V2/V3/V4 thing is still around
> it is just the contrary of the KISS ( Keep It Simple ) principle

V1 and V2 are needed to pass the Hercules version back to HercGUI.

> see my comment in dyngui.c

I did, thank you. The problem you describe there I was unaware of since I only work with Hyperion and not Spinhawk, so thank you for your comments. They are very much appreciated.

The problem your comments describe however can be easily address with a few lines of code in `makefile.bat`. My upcoming commit will illustrate this.

What to do on the Nix (non-Windows) side, I'm not sure.

Maybe an analogous `makefile.sh` script that does some up-front stuff (like try to build a meaningful VERSION string based on GIT repository information before invoking autogen, ./configure and make?

**Comment by srorso** [comment]
Sorry about that...at least I got the question right.  

> The GUI expects a floating point number of the form "number-dot-number" and not a string "a.b.c.d-gxxxx-modified". 

I missed this; apparently there's a communications protocol between HercGUI and dyngui.c, and version information in a specific form is needed to start a communications session. 

Is dyngui.c just there for HercGUI (making a "private" protocol fine), or does it implement a "standard" external panel interface that could be used to develop an alternative panel GUI, such as might be required in non-Windows environments?

Another idle question: could a HercGUI running in a Windows environment connect to a Hercules running on another perhaps virtualized non-Windows environment?

**Comment by Fish-Git** [comment]
> apparently there's a communications protocol between HercGUI and dyngui.c, and version information in a specific form is needed to start a communications session.

Actually no. HercGUI and dyngui are able to establish communication with one another right away without the need for the Hercules version information. The Hercules version information is simply for HercGUI's convenience so it can know whether or not the Hercules (dyngui) it is talking to supports a given "special HercGUI command" or not. It's not critical to HercGUI's functionality but it helps things.


> Is dyngui.c just there for HercGUI (making a "private" protocol fine), or does it implement a "standard" external panel interface that could be used to develop an alternative panel GUI, such as might be required in non-Windows environments?

There is no "standard" interface. It's just something I came up with myself years ago when HercGUI (and "External GUI Support" in Hercules) was first written and have extended and enhanced it over time.

As far as being used for alternative GUIs on non-Windows (or Windows!) platforms goes, it already _is_ being used!

Perhaps you were not aware that, as far as I know, there are at least two other Hercules GUIs for Linux, [Hercules Studio](http://www.jacobdekel.com/hercstudio/) and [Hebe](http://kde-apps.org/content/show.php/Hebe?content=126738)?

There is also another rather unique type of external GUI for Windows too, called [Jason](http://www.ollydbg.de/Jason/).

Someone suggested _long_ ago (_years and years_ ago in fact) about designing a more formal "communication protocol" for external GUI <---> Hercules communication, but due to workload (isn't it always that?) it never got off the ground. (There are always far too many things that need to be done and never enough hours in a day (or people!) to do them all in (which is why I really appreciate you joining the team!))


> Another idle question: could a HercGUI running in a Windows environment connect to a Hercules running on another perhaps virtualized non-Windows environment?

It's interesting that you should ask that because that is something I have been wanting to accomplish for a _very_ long time now: being able to control a Hercules running on Linux from HercGUI running on Windows.

And now that we have RFC compliant telnet support in Hercules this is something I intend to actively pursue in the not too distant future. &nbsp; `:)`

All of this however is **_terribly off topic_** to this particular GitHub Issue and should really be discussed elsewhere and not here. This GitHub Issue #102 is _supposed_ to just be about addressing the `V1`, `V2`, etc issue.

**Comment by Fish-Git** [comment]
This problem should now be resolved as of commit 91d24cf7e2c9142dd624d010ebde8f805f89e7fe.

Please feel free to re-open it if you discover any new issues/problems. Thank you.

(And again I would like to _THANK YOU_ Steve, for creating this issue and pointing out what the root cause actually was! Without your initial effort we might never have gotten VS2015 working with Hercules! So _thank_ for that!)

**Comment by srorso** [comment]
Happy to make a contribution...I'm learning and having  fun, so no worries, and thanks for the kind words.  I had to Google Union to figure out what it was--that's where my C skills are--and learned I've been using it for years in other languages.  I thought the risk of doing wrong was limited when I saw that the now-old code was in place for four years.  No matter; you have the vision for how that's supposed to work, and I'm pleased to let the future trump the past (a metaphor rooted in playing-card games, not U.S. politics.)

I'll be glad to move my OT content above to the Hercules-Devel; that's the right place for it. It would simplify future searches, even if it means everyone has to see the current posts twice.
